include/lbann/transforms/transform.hpp

```diff
@@ -0,0 +1,111 @@
+////////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2014-2016, Lawrence Livermore National Security, LLC.
+// Produced at the Lawrence Livermore National Laboratory.
+// Written by the LBANN Research Team (B. Van Essen, et al.) listed in
+// the CONTRIBUTORS file. <lbann-dev@llnl.gov>
+//
+// LLNL-CODE-697807.
+// All rights reserved.
+//
+// This file is part of LBANN: Livermore Big Artificial Neural Network
+// Toolkit. For details, see http://software.llnl.gov/LBANN or
+// https://github.com/LLNL/LBANN.
+//
+// Licensed under the Apache License, Version 2.0 (the "Licensee"); you
+// may not use this file except in compliance with the License.  You may
+// obtain a copy of the License at:
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+////////////////////////////////////////////////////////////////////////////////
+
+#ifndef LBANN_TRANSFORMS_TRANSFORM_HPP_INCLUDED
+#define LBANN_TRANSFORMS_TRANSFORM_HPP_INCLUDED
+
+#include "lbann/base.hpp"
+#include "lbann/utils/description.hpp"
+#include "lbann/utils/random.hpp"
+#include "lbann/utils/type_erased_matrix.hpp"
+#include "lbann/utils/exception.hpp"
+
+namespace lbann {
+namespace transform {
+
+/**
+ * Abstract base class for transforms on data.
+ * 
+ * A transform takes a CPUMat and modifies it in-place. Transforms should
+ * be thread-safe, as one instance of a transform may be called concurrently
+ * within multiple threads.
+ *
+ * Because transforms may switch between underlying data types throughout the
+ * pipeline, everything is done in terms of a type_erased_matrix, which can
+ * swap between underlying data types.
+ */
+class transform {
+public:
+  transform() = default;
+  transform(const transform&) = default;
+  transform& operator=(const transform&) = default;
+  virtual ~transform() = default;
+
+  /** Create a copy of the transform instance. */
+  virtual transform* copy() const = 0;
```

Comment:
In a world in which we return `unique_ptr<base>` from a `copy()`-like method (we should do this), the biggest issue comes from multiple inheritance (not necessarily the reality of multiple inheritance but also the mere possibility of multiple inheritance). The issue is unique to smart-pointer returns because raw-pointer returns can gloss over this detail with covariant return types but there's ambiguity with the smart-pointer version. The only thing, unfortunately, to fall back on (shy of some fun meta-magicks) is developer policy to name these something other than `copy()`, such as `copy_<interface>()`.